### DISCUSSION TITLE
sslmode=require not respected when using r2d2 pooler

### DISCUSSION THREAD

**Discussion by sjmiller609** [diesel-rs/diesel#3653] [Q&A] [answered]
### Versions

- **Rust:** rustc 1.67.0 (fc594f156 2023-01-24)
- **Diesel:** `diesel = { version = "2.1.0", features = ["postgres", "r2d2", "uuid", "chrono", "numeric"] }`
- **Database:**  postgres
- **Operating System** MacOS (applicable to other OS)

### Feature Flags

- **diesel:** `["postgres", "r2d2", "uuid", "chrono", "numeric"] `

## Problem Description

When providing sslmode=require in the query string, it's not respected and the client is behaving like sslmode=verify-ca

### What are you trying to accomplish?

Connecting to postgres server without TLS authenticating the server.

### What is the expected output?

With sslmode=require, I expect for the client to not check the CA, like it does with psql or other clients I have tested. That should be done with sslmode=verify-ca

Here is what the network looks like when it's working, this is with psql.

![image](https://github.com/diesel-rs/diesel/assets/7516283/37037417-9b14-446d-8377-cbeb261baac1)

### What is the actual output?

We see the TLS handshake completed, but the client is closing the connection with the error about verifying CA reported from client to server, then the client terminates the TCP connection.

```
thread 'main' panicked at 'Could not build connection pool: Error(Some("connection to server at \"org-cincinnati-ventures-inst-job-jar-staging.data-1.use1.coredb.io\" (44.209.131.194), port 5432 failed: SSL error: certificate verify failed\n"))', src/main.rs:23:14
```

![image](https://github.com/diesel-rs/diesel/assets/7516283/68be676e-958b-4451-990b-40ba5e1d4c9f)

### Are you seeing any additional errors?

No

### Steps to reproduce

This is a minimal reproduce setup: https://github.com/DarrenBaldwin07/coredb-gcp-issue/blob/main/src/main.rs

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/diesel-rs/diesel/issues) and the [discussion forum](https://github.com/diesel-rs/diesel/discussions) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by sjmiller609** [comment]
If we do not use the r2d2 pooler, it works fine, for example in this branch of the reproducing sample code.

https://github.com/DarrenBaldwin07/coredb-gcp-issue/blob/no-pooling/src/main.rs

**Comment by weiznich** [comment]
The r2d2 pool implementation does not change the provided database url as can bee seen here:

https://github.com/diesel-rs/diesel/blob/a4e6131b2457a43df216ecc17362857d796e3b47/diesel/src/r2d2.rs#L137-L142

and here:

https://github.com/diesel-rs/diesel/blob/a4e6131b2457a43df216ecc17362857d796e3b47/diesel/src/r2d2.rs#L191

This sounds like an environment issue for me, which is the reason why I converted this to a discussion

**Comment by sjmiller609** [comment, marked as answer]
I believe the reason for this issue is that I had this file present:

```
~/.postgresql/root.crt
```

So rust client was deciding that meant it should do verify-ca.

The reason my customer was having the issue was not because of CA verification and what they needed postgres version client version 14 or greater. Server Name Indication was introduced in version 14.
